pkg/cmd/cmd_update.go

```diff
@@ -48,39 +57,163 @@ func KpmUpdate(c *cli.Context, kpmcli *client.KpmClient) error {
+		if strings.Contains(pkgInfo, ":") {
+			return strings.Split(pkgInfo, ":")[0], strings.Split(pkgInfo, ":")[1]
 		}
+		return pkgInfo, ""
+	}
 
-		globalPkgPath, err := env.GetAbsPkgPath()
-		if err != nil {
-			return err
+	kclPkg, err := kpmcli.LoadPkgFromPath(pwd)
+	if err != nil {
+		return err
+	}
+
+	globalPkgPath, err := env.GetAbsPkgPath()
+	if err != nil {
+		return err
+	}
+
+	err = kclPkg.ValidateKpmHome(globalPkgPath)
+	if err != (*reporter.KpmEvent)(nil) {
+		return err
+	}
+
+	var (
+		modulesToUpgrade   []module.Version
+		modulesToDowngrade []module.Version
+	)
+
+	// validate if the packages is present in kcl.mod file and
+	// find the latest version if version is not specified
+	for _, pkgInfo := range pkgInfos {
+		pkgInfo = strings.TrimSpace(pkgInfo)
+		pkgName, pkgVersion := fetchNameAndVersion(pkgInfo)
+
+		var dep pkg.Dependency
+		var ok bool
+		if dep, ok = kclPkg.Deps[pkgName]; !ok {
+			reporter.Report(fmt.Sprintf("package %s not found in kcl.mod file", pkgName))
+			continue
 		}
 
-		err = kclPkg.ValidateKpmHome(globalPkgPath)
-		if err != (*reporter.KpmEvent)(nil) {
-			return err
+		if pkgVersion == "" {
+			var releases []string
+			if dep.Git != nil {
+				releases, err = mvs.GetReleasesFromSource(dep.GetSourceType(), dep.Git.Url)
+			} else if dep.Oci != nil {
+				releases, err = mvs.GetReleasesFromSource(dep.GetSourceType(), "fjdkasd")
+			}
+			if err != nil {
+				reporter.ReportEventToStdout(reporter.NewEvent(
+					reporter.FailedGetReleases,
+					fmt.Sprintf("failed to get releases for %s", pkgName),
+				))
+				continue
+			}
+			pkgVersion, err = semver.LatestCompatibleVersion(releases, dep.Version)
+			if err != nil {
+				reporter.NewErrorEvent(
+					reporter.FailedSelectLatestCompatibleVersion,
+					err,
+					fmt.Sprintf("failed to find the latest version for %s", pkgName),
+				)
+				continue
+			}
+		}
+		if pkgVersion < dep.Version {
+			modulesToDowngrade = append(modulesToDowngrade, module.Version{Path: pkgName, Version: pkgVersion})
+		} else if pkgVersion > dep.Version {
+			modulesToUpgrade = append(modulesToUpgrade, module.Version{Path: pkgName, Version: pkgVersion})
 		}
+	}
+
+	_, depGraph, err := kpmcli.InitGraphAndDownloadDeps(kclPkg)
+	if err != nil {
+		return err
+	}
+
+	reqs := mvs.ReqsGraph{
+		Graph:     depGraph,
+		KpmClient: kpmcli,
+		KpmPkg:    kclPkg,
+	}
+
+	target := module.Version{Path: kclPkg.GetPkgName(), Version: kclPkg.GetPkgVersion()}
+	buildList, err := UpdateBuildList(target, modulesToUpgrade, modulesToDowngrade, &reqs)
+	if err != nil {
+		return reporter.NewErrorEvent(reporter.FailedUpdatingBuildList, err, "failed to update build list")
+	}
+
+	modules, err := graph.TopologicalSort(depGraph)
+	if err != nil {
+		return reporter.NewErrorEvent(reporter.FailedTopologicalSort, err, "failed to sort the dependencies")
+	}
 
-		err = kpmcli.UpdateDeps(kclPkg)
+	kclPkg.ModFile.Dependencies.Deps = make(map[string]pkg.Dependency)
```

Comment:
This function is a bit long, you can split it into smaller functions.

Reply:
done